### ISSUE TITLE
May I ask why you choose the index 0 token of groundtruth when calculating the MI in the file cal_mi.py?

### ISSUE THREAD

**Issue by yokiwuuu** [ChnQ/MI-Peaks#4]
```

for layer in layers:  
            here_num_tokens = acts[id]['reps'][layer].shape[0]  
            layer_mi_list = torch.zeros(here_num_tokens)
            for i in range(here_num_tokens):
                layer_mi_list[i] =  estimate_mi_hsic(acts[id]['reps'][layer][i], gt_acts[id][layer][0])
                print(f'loop {i} finished!')

            layer_mi_matrix.append(layer_mi_list)
            final_mi_dict[id]['reps'][layer] = layer_mi_list  
```

**Could you explain more about  the meaning of this index 0 token in gt_acts?**

**Comment by ChnQ** [comment]
Thank you for your interest!

Since we only save the representation of the last token when [generating the gt representations](https://github.com/ChnQ/MI-Peaks/blob/master/src/generate_gt_activation.py), `gt_acts[id][layer]` actually contains only that one element.

**Comment by yokiwuuu** [comment]
Thank you for your reply. So the last token of gt is the final answer? Did you consider the whole reasoning solution as the gt?

**Comment by ChnQ** [comment]
We choose to use the representation of the last token in the gt answer to represent the entire answer because previous studies [1-5, etc.] has shown that the last token representation often contains the information of the whole sentence.

Of course, I believe it’s also reasonable to explicitly “consider the whole reasoning solution”—for example, by averaging the representations of all tokens.

[1] https://huggingface.co/docs/transformers/model_doc/llama
[2] Li, Zongxi, et al. "Label supervised llama finetuning." arXiv preprint arXiv:2310.01208 (2023).
[3] Jiang, Ting, et al. "Scaling sentence embeddings with large language models." ACL 2024.
[4] Li, Kenneth, et al. "Inference-time intervention: Eliciting truthful answers from a language model." NeurIPS 2023.
[5] Gurnee, Wes, and Max Tegmark. "Language models represent space and time." ICLR 2024.
